enhancements/ingress/aws-load-balancer-operator.md

```diff
@@ -0,0 +1,384 @@
+  ingressClass: alb
+  rules:
+  - host: example.com
+    http:
+        paths:
+- path: /store
+  backend:
+    service:
+name: example-2
+port:
+  number: 80
+—--
+kind: Ingress
+metadata:
+  name: example-3
+  annotations:
+    alb.ingress.kubernetes.io/scheme: internet-facing
+    alb.ingress.kubernetes.io/group.name: example
+    alb.ingress.kubernetes.io/group.order: “3”
+spec:
+  ingressClass: alb
+  rules:
+  - host: example.com
+    http:
+        paths:
+- path: /
+  backend:
+    service:
+name: example-3
+port:
+  number: 80
+```
+
+In this example the annotations `alb.ingress.kubernetes.io/group.name` specifies
+which Ingress Group the Ingresses belong to and the annotation
+`alb.ingress.kubernetes.io/group.order` specifies the order in the group which
+will be used for path matching. All Ingresses which belong to the same group are
+attached to the same ALB.
+
+### API Extensions
+
+This operator uses a cluster-scoped Custom Resource to configure the
+lb-controller instance. The resource has the following _Spec_:
+
+```golang
+type AWSLoadBalancerControllerSpec struct {
+  AutoTagSubnet bool // indicates if the operator should tag the subnets
+  // +optional
+  AdditionalResourceTags map[string]string // Default AWS Tags that will be applied to all AWS resources managed by this controller (default [])
+  // +optional
+  IngressClass string   // indicates the class of ingress for which the controller should provision ALBs
+  // +optional
+  Config *DeploymentConfig
+  // +optional
+  DisabledAddons []AWSAddon // indicates which AWS addons should be disabled.
+}
+
+type AWSAddon string
+
+const (
+  AWSAddonShield AWSAddon = “AWSShield”
+  AWSAddonWAFv1 AWSAddon = “AWSWAFv1”
+  AWSAddonWAFv2 AWSAddon = “AWSWAFv2”
+)
+
+type DeploymentConfig struct {
+  // +optional
+  Replicas int
+}
+```
+
+
+### Implementation Details/Notes/Constraints
+
+The operator only reconciles a single instance of the custom resource because
+only [one
+instance](https://github.com/kubernetes-sigs/aws-load-balancer-controller/issues/2185)
+of the lb-controller can be run in a cluster. There can be multiple replicas
+with the same configuration with leader-election enabled. But multiple
+controllers cannot be started with different configurations.
+
+The lb-controller also requires that the subnets where the load balancers are
+provisioned have certain resource tags present on them. The operator can detect
+the subnets and tag them or the user can do this. This is specified with the
+_AutoTagSubnet_ field.
+
+The user can also specify the Ingress class which the lb-controller will
+reconcile. This could default to **_“alb”_**. This value is required because if
+an ingress class value is not passed to the controller it attempts to reconcile
+all Ingress resources which don’t have an ingress class annotation or where the
+ingress class field is empty. Passing the value of “alb” does not restrict the
+number of ingress classes which the controller can reconcile.
```

Comment:
> does not restrict the number of ingress classes which the controller can reconcile.

Didn't quite get this.
You meant the number of ingress groups (the one you can set with `group.name` annotation)?

Reply:
No I mean IngressClasses themselves. If the `IngressClass` has the `spec.controller` set to `ingress.k8s.aws/alb` then the controller will reconcile those Ingresses regardless of the ingress class passed to it through the CLI argument. You can find more details in the "Parallel Operation" section.